### PR TITLE
Closes #14 as_slice returns the &[u8] slice implied by offset and len.

### LINKED ISSUE #14: MmapReadOnly::as_slice() does not take in account offset and len
`as_slice()` returns the whole byte array.

Should not it take in account `offset` and `len`?

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

File: src/raw/mod.rs [hunk 1]
```diff
@@ -937,9 +938,14 @@ impl Output {
     }
 }
 
-enum FstData {
-    Cow(Cow<'static, [u8]>),
+/// Represent the data source for an FST>
+pub enum FstData {
```
Comment (BurntSushi) [inline src/raw/mod.rs]: I don't think this should be public. Why do you need it?

File: src/raw/mmap.rs [hunk 2]
```diff
@@ -53,7 +53,7 @@ impl MmapReadOnly {
 
     /// Read the memory map as a `&[u8]`.
     pub unsafe fn as_slice(&self) -> &[u8] {
-        self.map.as_slice()
+        &self.map.as_slice()[self.offset..self.offset + self.len]
```
Comment (BurntSushi) [inline src/raw/mmap.rs]: Good catch!

File: src/raw/shared_vector_slice.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        assert!(offset + len <= self.len);
+        SharedVectorSlice {
+            bytes: self.bytes.clone(),
+            offset: self.offset + offset,
+            len: len,
+        }
+    }
+
+    /// Read the memory map as a `&[u8]`.
+    pub unsafe fn as_slice(&self) -> &[u8] {
```
Comment (BurntSushi) [inline src/raw/shared_vector_slice.rs]: I think this is safe. You can remove the `unsafe` annotation.

File: src/raw/mod.rs [hunk 4]
```diff
@@ -947,8 +953,9 @@ impl Deref for FstData {
 
     fn deref(&self) -> &[u8] {
         match *self {
-            FstData::Cow(ref v) => &**v,
+            FstData::Static(ref v) => v,
             FstData::Mmap(ref v) => unsafe { v.as_slice() },
+            FstData::SharedVector(ref v) => unsafe { v.as_slice() },
```
Comment (BurntSushi) [inline src/raw/mod.rs]: Once you remove the `unsafe` annotation, you should be able to elide it here too.

**Comment by BurntSushi** [comment]
OK, so I think the issue here is that this doesn't actually solve your problem. In order to solve your problem, I think you need to make `SharedVectorSlice` public (unfortunately). Then you'll need a new constructor function on `Fst` that takes a `SharedVectorSlice`.

File: src/raw/mmap.rs (hunk 2 above)
Comment (fulmicoton) [inline src/raw/mmap.rs]: Ouch sorry ... I think github automatically added extra commit to a PR if it is on the same branch.  These should be two separate MR.
